From 189471fd6c4d59db77a1195d1e2b36551ed63134 Mon Sep 17 00:00:00 2001
From: Ping Gao <ping.a.gao@intel.com>
Date: Tue, 26 Jan 2016 22:08:32 +0800
Subject: [PATCH 323/403] enable shadow indirect ctx and batch buffer per ctx

set the module parameter to true and replace the corresponding mmio
value in shadow ctx by shadow value.

Signed-off-by: Ping Gao <ping.a.gao@intel.com>
---
 drivers/gpu/drm/i915/vgt/execlists.c |   11 +++++++++++
 drivers/gpu/drm/i915/vgt/vgt.c       |    2 +-
 2 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/vgt/execlists.c b/drivers/gpu/drm/i915/vgt/execlists.c
index 1fb7ef6..767d451 100644
--- a/drivers/gpu/drm/i915/vgt/execlists.c
+++ b/drivers/gpu/drm/i915/vgt/execlists.c
@@ -730,6 +730,17 @@ static void update_shadow_regstate_from_guest(struct vgt_device *vgt,
 	if (shadow_cmd_buffer)
 		dest_ctx->rb_start.val = el_ctx->shadow_rb.shadow_rb_base;
 
+	if (shadow_indirect_ctx_bb) {
+		dest_ctx->rcs_indirect_ctx.val =
+			(dest_ctx->rcs_indirect_ctx.val &
+				(~INDIRECT_CTX_ADDR_MASK)) |
+				el_ctx->shadow_indirect_ctx.shadow_ctx_base;
+		dest_ctx->bb_per_ctx_ptr.val =
+			(dest_ctx->bb_per_ctx_ptr.val &
+				(~BB_PER_CTX_ADDR_MASK)) |
+				el_ctx->shadow_bb_per_ctx.shadow_bb_base;
+	}
+
 	ppgtt_update_shadow_ppgtt_for_ctx(vgt, el_ctx);
 }
 
diff --git a/drivers/gpu/drm/i915/vgt/vgt.c b/drivers/gpu/drm/i915/vgt/vgt.c
index 504d42d..ac72e9f 100644
--- a/drivers/gpu/drm/i915/vgt/vgt.c
+++ b/drivers/gpu/drm/i915/vgt/vgt.c
@@ -194,7 +194,7 @@ module_param_named(shadow_cmd_buffer, shadow_cmd_buffer, int, 0400);
 int shadow_ctx_check = 0;
 module_param_named(shadow_ctx_check, shadow_ctx_check, int, 0600);
 
-int shadow_indirect_ctx_bb = 0;
+int shadow_indirect_ctx_bb = 1;
 module_param_named(shadow_indirect_ctx_bb, shadow_indirect_ctx_bb, int, 0400);
 
 static struct vgt_ops __vgt_ops = {
-- 
1.7.10.4

